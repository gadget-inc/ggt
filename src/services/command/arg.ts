import arg from "arg";
import type { Simplify } from "type-fest";
import { GGTError, IsBug, UnexpectedError } from "../output/report.js";
import { isNil } from "../util/is.js";

export type ArgsDefinition = Record<string, ArgDefinition>;

type ArgDefinition<Handler extends arg.Handler = arg.Handler> =
  | Handler
  | {
      type: Handler;
      alias?: string | string[];
      default?: ReturnType<Handler>;
    };

export type ParseArgsOptions = {
  /**
   * A list of arguments to parse.
   */
  argv?: string[];

  /**
   * When permissive set to `true`, arg will push any unknown arguments
   * onto the "extra" argument array (`ctx.args._`) instead of throwing
   * an error about an unknown flag.
   *
   * @default false
   */
  permissive?: boolean;

  /**
   * When stopAtPositional is set to true, context will stop parsing at
   * the first positional argument.
   *
   * @default false
   */
  stopAtPositional?: boolean;
};

export const parseArgs = <Args extends ArgsDefinition>(args: Args, options?: arg.Options): ArgsDefinitionResult<Args> => {
  const spec: arg.Spec = {};
  const defaultValues: Record<string, unknown> = {};

  for (const [key, value] of Object.entries(args)) {
    if (!("type" in value)) {
      spec[key] = value;
      continue;
    }

    spec[key] = value.type;
    defaultValues[key] = value.default;

    if (value.alias) {
      for (const alias of Array.isArray(value.alias) ? value.alias : [value.alias]) {
        spec[alias] = key;
      }
    }
  }

  try {
    const parsed = arg(spec, options);
    for (const [key, value] of Object.entries(defaultValues)) {
      if (isNil(parsed[key])) {
        parsed[key] = value as never;
      }
    }
    return parsed as ArgsDefinitionResult<Args>;
  } catch (error: unknown) {
    if (error instanceof arg.ArgError) {
      // convert arg.ArgError to GGTError
      // eslint-disable-next-line no-ex-assign
      error = new ArgError(error.message);
    }
    if (error instanceof GGTError) {
      throw error;
    }
    throw new UnexpectedError(error);
  }
};

export class ArgError extends GGTError {
  isBug = IsBug.NO;

  protected override render(): string {
    return this.message;
  }
}

/**
 * Turns this:
 * ```ts
 * type Args = {
 *   "--string": { type: String; alias: "s" };
 *   "--number": { type: Number; default: 42 };
 * };
 * ```
 *
 * Into this:
 * ```ts
 * type Result = {
 *  "--string": string | undefined;
 *  "--number": number;
 * };
 * ```
 */
export type ArgsDefinitionResult<Args extends ArgsDefinition, Keys extends keyof Args = keyof Args> = Simplify<{
  [Key in Keys]: Args[Key] extends ArgDefinition<infer Handler>
    ? Args[Key] extends { default: unknown }
      ? NonNullable<ReturnType<Handler>>
      : ReturnType<Handler> | undefined
    : never;
}> & { _: string[] };
