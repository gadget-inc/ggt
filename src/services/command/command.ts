import assert from "node:assert";
import { pathToFileURL } from "node:url";
import type { EmptyObject, Promisable } from "type-fest";
import type { RootArgs } from "../../commands/root.js";
import { config } from "../config/config.js";
import { relativeToThisFile } from "../util/paths.js";
import type { ArgsDefinition } from "./arg.js";
import type { Context } from "./context.js";

/**
 * The list of available commands.
 *
 * 1. Every command corresponds to a file inside of src/commands/
 * 2. The order determines the order of commands in the README
 */
export const Commands = ["dev", "deploy", "status", "push", "pull", "add", "open", "list", "login", "logout", "whoami", "version"] as const;

/**
 * One of the commands in {@link Commands}.
 */
export type AvailableCommand = (typeof Commands)[number];

/**
 * Checks if a string is a valid command.
 *
 * @param command - The string to check
 * @returns Whether the string is a valid command
 */
export const isAvailableCommand = (command: string): command is AvailableCommand => {
  return Commands.includes(command as AvailableCommand);
};

/**
 * A command module is a file in the src/commands/ directory.
 */
export type CommandModule<Args extends ArgsDefinition = EmptyObject, ParentArgs extends ArgsDefinition = RootArgs> = {
  /**
   * The command's {@link ArgsDefinition args}.
   */
  args?: Args;

  /**
   * The command's {@link Usage usage}.
   */
  usage: Usage;

  /**
   * The command's {@link Command command}.
   *
   * TODO: rename this to `run`.
   */
  command: Command<Args, ParentArgs>;
};

/**
 * A {@linkcode Command command}'s usage is a function that returns a
 * string describing how to use the command. The function receives its
 * parent command's context.
 */
export type Usage = (ctx: Context) => string;

/**
 * The function that is run when the command is called.
 *
 * @param ctx - A {@linkcode Context} with the command's {@linkcode Args} and {@linkcode ParentArgs}.
 */
export type Command<Args extends ArgsDefinition = EmptyObject, ParentArgs extends ArgsDefinition = RootArgs> = (
  ctx: Context<Args, ParentArgs>,
) => Promisable<void>;

/**
 * Imports a command module.
 *
 * @param cmd - The command to import
 * @see {@linkcode CommandModule}
 */
export const importCommand = async (cmd: AvailableCommand): Promise<CommandModule> => {
  assert(isAvailableCommand(cmd), `invalid command: ${cmd}`);
  let commandPath = relativeToThisFile(`../../commands/${cmd}.js`);
  if (config.windows) {
    // https://github.com/nodejs/node/issues/31710
    commandPath = pathToFileURL(commandPath).toString();
  }
  return (await import(commandPath)) as CommandModule;
};
