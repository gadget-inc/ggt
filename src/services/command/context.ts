import { createLogger, type Logger } from "../output/log/logger.js";
import type { StructuredLoggerOptions } from "../output/log/structured.js";
import { PromiseSignal } from "../util/promise.js";
import type { AnyVoid } from "../util/types.js";

/**
 * Represents the context of a command-line operation.
 */
export class Context extends AbortController {
  /**
   * A promise that resolves when the context is aborted and all the
   * registered onAbort callbacks have finished.
   */
  readonly done = new PromiseSignal<void>();

  /**
   * The logger for the current context.
   */
  #log: Logger;

  /**
   * The values that have been set on this context.
   */
  #values: Record<symbol, unknown>;

  /**
   * The callbacks that will be called when this context is aborted.
   */
  #onAborts: OnAbort[] = [];

  private constructor({ log, values }: { log: Logger; values: Record<symbol, unknown> }) {
    super();
    this.#log = log;
    this.#values = values;

    // in case this context is ...spread into another object
    this.abort = this.abort.bind(this);
    this.child = this.child.bind(this);
    this.onAbort = this.onAbort.bind(this);

    // when the context is aborted, call all the registered callbacks
    this.signal.addEventListener(
      "abort",
      // eslint-disable-next-line @typescript-eslint/no-misused-promises
      async () => {
        let error: unknown;

        // call the callbacks in reverse order, like go's defer
        for (const callback of this.#onAborts.reverse()) {
          try {
            await callback(this.signal.reason);
          } catch (e: unknown) {
            error ??= e;
            this.log.error("error during abort", { error });
          }
        }

        if (error) {
          this.done.reject(error);
        } else {
          this.done.resolve();
        }
      },
    );
  }

  /**
   * A {@linkcode Logger} that can print to stdout and log structured
   * messages to stderr.
   */
  get log(): Logger {
    return this.#log;
  }

  /**
   * Initializes a new context.
   */
  static init(options: StructuredLoggerOptions): Context {
    return new Context({
      log: createLogger(options),
      values: {},
    });
  }

  get(key: symbol): unknown {
    return this.#values[key];
  }

  set(key: symbol, value: unknown): void {
    this.#values[key] = value;
  }

  /**
   * Returns a new context that is a child of the current context.
   *
   * @see {@linkcode ContextInit}
   */
  child(options: StructuredLoggerOptions): Context {
    const ctx = new Context({
      log: this.log.child(options),
      values: this.#values,
    });

    this.onAbort(() => ctx.abort());

    return ctx;
  }

  /**
   * Registers a callback that will be called when the context is
   * aborted (e.g. when the user presses Ctrl+C).
   *
   * @param callback - The callback to call when the context is aborted.
   */
  onAbort(callback: OnAbort): void {
    this.#onAborts.push(callback);
  }
}

/**
 * A callback that will be called when the context is aborted.
 */
export type OnAbort = (reason: unknown) => AnyVoid;
