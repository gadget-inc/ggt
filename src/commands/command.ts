import arg from "arg";
import assert from "node:assert";
import type { Context } from "src/services/command/context.js";
import { relativeToThisFile } from "src/services/config/paths.js";
import type { Promisable } from "type-fest";

export type Usage = () => string;

export type Command = (ctx: Context) => Promisable<void>;

export type CommandModule = {
  usage: Usage;
  command: Command;
};

export const AvailableCommands = ["sync", "list", "login", "logout", "whoami", "version", "deploy"] as const;

type AvailableCommand = (typeof AvailableCommands)[number];

export const isAvailableCommand = (value: unknown): value is AvailableCommand => {
  return AvailableCommands.includes(value as AvailableCommand);
};

export const importCommandModule = async (command: AvailableCommand): Promise<CommandModule> => {
  assert(isAvailableCommand(command), `invalid command: ${command}`);
  return (await import(relativeToThisFile(`../../commands/${command}.js`))) as CommandModule;
};

export const rootArgsSpec = {
  "--help": Boolean,
  "-h": "--help",
  "--verbose": arg.COUNT,
  "-v": "--verbose",
  "--json": Boolean,

  // deprecated
  "--debug": "--verbose",
};

export type RootArgs = arg.Result<typeof rootArgsSpec>;
