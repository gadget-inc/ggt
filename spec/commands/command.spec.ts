import { beforeAll, describe, expect, it } from "vitest";
import { AvailableCommands, importCommandModule, type CommandModule } from "../../src/commands/command.js";

describe.each(AvailableCommands)("%s", (command) => {
  let mod: CommandModule;

  beforeAll(async () => {
    mod = await importCommandModule(command);
  });

  it("has a usage", () => {
    expect(mod.usage).toBeDefined();
    expect(mod.usage).toBeInstanceOf(Function);
    expect(mod.usage()).toMatchSnapshot();
  });

  it("has a command", () => {
    expect(mod.command).toBeDefined();
    expect(mod.command).toBeInstanceOf(Function);
  });
});
