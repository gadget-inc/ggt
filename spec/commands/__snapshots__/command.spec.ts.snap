// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`list > has a usage 1`] = `
"List the apps available to the currently logged in user.

USAGE
  ggt list

EXAMPLE
  $ ggt list
    Slug    Domain
    ─────── ──────────────────
    my-app  my-app.gadget.app
    example example.gadget.app
    test    test.gadget.app"
`;

exports[`login > has a usage 1`] = `
"Log in to your account.

USAGE
  ggt login

EXAMPLES
  $ ggt login
    We've opened Gadget's login page using your default browser.

    Please log in and then return to this terminal.

    Hello, Jane Doe (jane@example.com)"
`;

exports[`logout > has a usage 1`] = `
"Log out of your account.

USAGE
  ggt logout

EXAMPLES
  $ ggt logout
    Goodbye"
`;

exports[`sync > has a usage 1`] = `
"Sync your Gadget application's source code to and from
your local filesystem.

USAGE
  ggt sync [DIRECTORY] [--app=<name>]

ARGUMENTS
  DIRECTORY         The directory to sync files to (default: \\".\\")

FLAGS
  -a, --app=<name>  The Gadget application to sync files to
      --force       Sync regardless of local file state

DESCRIPTION
  Sync allows you to synchronize your Gadget application's source
  code with your local filesystem.

  While ggt sync is running, local file changes are immediately
  reflected within Gadget, while files that are changed in Gadget are
  immediately saved to your local filesystem.

  Ideal for:
    • Local development with editors like VSCode
    • Storing source code in a Git repository like GitHub

  Sync looks for a \\".ignore\\" file to exclude certain files/directories
  from being synced. The format is identical to Git's.

  These files are always ignored:
    • .DS_Store
    • .gadget
    • .git
    • node_modules

  Note:
    • Sync only works with your development environment
    • Avoid deleting/moving all your files while sync is running
    • Gadget only supports Yarn v1 for dependency installation

EXAMPLE
  $ ggt sync ~/gadget/example --app example

    App         example
    Editor      https://example.gadget.app/edit
    Playground  https://example.gadget.app/api/graphql/playground
    Docs        https://docs.gadget.dev/api/example

    Endpoints
      • https://example.gadget.app
      • https://example--development.gadget.app

    Watching for file changes... Press Ctrl+C to stop

    → Sent 09:06:25 AM
    routes/GET-hello.js  + create

    → Sent 09:06:49 AM
    routes/GET-hello.js  ± update

    ← Received 09:06:54 AM
    routes/GET-hello.js  ± update

    ← Received 09:06:56 AM
    routes/GET-hello.js  - delete
    ^C Stopping... press Ctrl+C again to force

    Goodbye!"
`;

exports[`version > has a usage 1`] = `
"Print the version of ggt

USAGE
  ggt version

EXAMPLES
  $ ggt version
    0.3.3"
`;

exports[`whoami > has a usage 1`] = `
"Show the name and email address of the currently logged in user

USAGE
  ggt whoami

EXAMPLES
  $ ggt whoami
    You are logged in as Jane Doe (jane@example.com)"
`;
